wdl-ast/src/v1/validation/names.rs

```diff
@@ -0,0 +1,237 @@
+//! Validation of unique names in a V1 AST.
+
+use std::collections::HashMap;
+use std::fmt;
+
+use wdl_grammar::Diagnostic;
+use wdl_grammar::Span;
+
+use crate::v1::BoundDecl;
+use crate::v1::ImportStatement;
+use crate::v1::ScatterStatement;
+use crate::v1::StructDefinition;
+use crate::v1::TaskDefinition;
+use crate::v1::UnboundDecl;
+use crate::v1::Visitor;
+use crate::v1::WorkflowDefinition;
+use crate::AstToken;
+use crate::Diagnostics;
+use crate::Ident;
+use crate::VisitReason;
+
+/// Represents context about a unique name validation error.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum Context {
+    /// The error is a task name.
+    Task,
+    /// The error is a struct name.
+    Struct,
+    /// The error is a struct member name.
+    StructMember,
+    /// The error is a declaration name.
+    Declaration,
+}
+
+impl fmt::Display for Context {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Task => write!(f, "task"),
+            Self::Struct => write!(f, "struct"),
+            Self::StructMember => write!(f, "struct member"),
+            Self::Declaration => write!(f, "declaration"),
+        }
+    }
+}
+
+/// Creates a "name conflict" diagnostic
+fn name_conflict(context: Context, name: Ident, first: Span) -> Diagnostic {
+    Diagnostic::error(format!(
+        "conflicting {context} name `{name}`",
+        name = name.as_str(),
+    ))
+    .with_label(
+        format!("this conflicts with a previous {context} of the same name"),
+        name.span(),
+    )
+    .with_label(format!("first {context} with this name is here"), first)
+}
+
+/// A visitor of unique names within an AST.
+///
+/// Ensures that the following names are unique:
+///
+/// * Task names.
+/// * Struct names from struct declarations and import aliases.
+/// * Struct member names.
+/// * Declarations and scatter variable names.
+///
+/// Note that it does not check for duplicate workflow names as it's already a
+/// validation error to have more than one workflow in a document.
```

Comment:
Isn't it a problem to have a task and workflow with the same name? e.g.
```wdl
version 1.1
task foo {}
workflow foo {}
```
If that's in `foo.wdl` and in `bar.wdl` I have
```wdl
version 1.1
import foo.wdl

workflow bar { call foo.foo }
```
Does the second `foo` call the workflow or the task?

Reply:
Given the example to the `name_collision` lint in the baseline lint rules discussion, it sounded like the names only had to be unique for their "kind"; the spec also isn't very clear on it either.

For what it's worth, `miniwdl` does check that the name is unique for a top-level item in the document (including struct names); I personally think that what `miniwdl` is checking makes the most sense and I was quite confused when I first read the description of the proposed `name_collision` rule.

Certainly the example you have there does make it seem like that would be ambiguous to an execution engine and require that the task and workflow have unique names.

Reply:
Also, I realized that we should be validating the namespace names being introduced by import statements as being unique as well (either the implicit or explicit name).

If there's no explicit name and it conflicts, then we can suggest using an `as` clause on the import in the diagnostic.

Reply:
> Given the example to the name_collision lint in the baseline lint rules discussion, it sounded like the names only had to be unique for their "kind"; the spec also isn't very clear on it either.

Yes, I just reviewed the example I gave in that document and it is incorrect. Whoops 😬. 

> For what it's worth, miniwdl does check that the name is unique for a top-level item in the document (including struct names);

So as we established above, a declared workflow must have a unique name amongst any declared tasks. But I think `miniwdl` is going a step beyond what is required by the SPEC. I think a struct name should be able to overlap with a workflow or task. (It should definitely be advised against, but should be syntactically clear to my understanding).

> Also, I realized that we should be validating the namespace names being introduced by import statements as being unique as well (either the implicit or explicit name).
>
> If there's no explicit name and it conflicts, then we can suggest using an as clause on the import in the diagnostic.

Sounds good 👍

Reply:
So having a task and workflow with the same name does occur in the wild:

https://github.com/broadinstitute/gatk-tool-wdls/blob/master/wdls/ApplyBQSR.wdl

Reply:
Unresolving because I have realized _another_ issue with my prior example (😬 I guess it was a _really_ bad example). Rewriting here for ease of reference:

`foo.wdl`
```wdl
version 1.1
task foo {}
workflow foo {}
```
`bar.wdl`
```wdl
version 1.1
import foo.wdl
workflow bar { call foo.foo }
```
So we've talked about the name collision on `foo` _within_ `foo.wdl` (exemplified by the ambiguous `call` in `bar.wdl`), and you've addressed that and the change looks great 👍 .

BUT, there is another name collision! In `bar.wdl` L2 introduces the namespace-`foo` into the scoped-namespace. Then in L3 `call foo.foo` tries to introduce the call-name-`foo` into the same scoped-namespace! This would have to be fixed in one of two ways (let's pretend `foo` resolves unambiguously):
```wdl
version 1.1
import foo.wdl as something_else
workflow bar { call something_else.foo }
```
OR
```wdl
version 1.1
import foo.wdl
workflow bar { call foo.foo as something_else }
```
Note that we should prefer the first option (it's easier to track aliases when they're in the import block than when they're in a workflow body).

Reply:
> Unresolving because I have realized _another_ issue with my prior example (😬 I guess it was a _really_ bad example). Rewriting here for ease of reference:
> 
> `foo.wdl`
> 
> ```wdl
> version 1.1
> task foo {}
> workflow foo {}
> ```
> 
> `bar.wdl`
> 
> ```wdl
> version 1.1
> import foo.wdl
> workflow bar { call foo.foo }
> ```
> 
> So we've talked about the name collision on `foo` _within_ `foo.wdl` (exemplified by the ambiguous `call` in `bar.wdl`), and you've addressed that and the change looks great 👍 .
> 
> BUT, there is another name collision! In `bar.wdl` L2 introduces the namespace-`foo` into the scoped-namespace. Then in L3 `call foo.foo` tries to introduce the call-name-`foo` into the same scoped-namespace! This would have to be fixed in one of two ways (let's pretend `foo` resolves unambiguously):
> 
> ```wdl
> version 1.1
> import foo.wdl as foo_namespace
> workflow bar { call foo_namespace.foo }
> ```
> 
> OR
> 
> ```wdl
> version 1.1
> import foo.wdl
> workflow bar { call foo.foo as foo_call }
> ```
> 
> Note that we should prefer the first option (it's easier to track aliases when they're in the import block than when they're in a workflow body).

I asked @adthrasher to double check because I didn't feel confident about this, and I am wrong yet again 😬 Apparently I am really bad with conceptualizing namespaces lol.
`bar.wdl` is fine, it does have multiple things in scope named `foo` but they are accessed in different contexts and therefore they can be referred to unambiguously.
* `foo`: introduced by `import` is a task/workflow namespace
* `foo` introduced by `call foo.foo` is a declaration namespace

Those name collisions should probably be lint warnings, but they are ok in the context of this PR (validation-wise).

Reply:
@a-frantz and I discussed this this morning on Slack and walked through some examples. My opinion is that we should be overzealous in using lint warnings to discourage name conflicts (even if permissible in the spec/execution engine). After all, we are attempting to create rules to encourage well-formed, readable WDL files. Name conflicts are difficult for a human reader to decipher.